[master] rest_cherrypy: accept / force credentials via HTTP header

Malte-Wagner opened the pull request:
The acceptance of HTTP headers opens the possibility to either set or force credentials for a session.
A reverse proxy, can handle authentication and
inject the custom headers X-Forwarded-User, X-Forwarded-Password and/or X-Forwarded-Eauth. In general, these headers can work in conjunction with the eauth sharedsecret / rest / auto methods to externalize and flexibilize authentication, without raising code complexity for new or niche authentication methods inside salt.

### What does this PR do?
This PR enables rest_cherrypy to gather credentials via HTTP headers, that may be set by reverse proxy, to use them for session creation or session checking.

### What issues does this PR fix or reference?
https://github.com/saltstack/salt/issues/22046

### Previous Behavior
Credentials are only allowed via request body.

### New Behavior
Credentials are allowed via request body and http header.
Prevalence is http header over request body values.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [x] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices, including the
[PR Guidelines](https://docs.saltproject.io/en/master/topics/development/pull_requests.html).

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

Malte-Wagner committed:
rest_cherrypy: accept / force credentials via HTTP header

The acceptance of HTTP headers opens the possibility to either set
or force credentials for a session.
A reverse proxy, can handle authentication and
inject the custom headers X-Forwarded-User, X-Forwarded-Password
and/or X-Forwarded-Eauth. In general, these headers can work in
conjunction with the eauth sharedsecret / rest / auto methods to
externalize and flexibilize authentication, without raising
code complexity for new or niche authentication methods inside salt.

twangboy commented on salt/netapi/rest_cherrypy/app.py:
```diff
@@ -240,9 +240,26 @@
+
+### Session start and check via custom header
+   If one or more custom headers are present ``X-Forwarded-User``,
+   ``X-Forwarded-User`` or ``X-Forwarded-Eauth`` the value will **override** the
```
X-Forwarded-User is in this list twice, maybe missing X-Forwarded-Password?

twangboy reviewed (changes requested):
Also needs a changelog and some tests, please.

dwoz committed:
Fix review issues: changelog, tests, and doc/code cleanup

- Fix docstring bug: X-Forwarded-User listed twice instead of X-Forwarded-Password
- Fix RST heading format (### is not valid RST; use underline style)
- Fix typos: "diffent" -> "different", grammar in comment
- Fix trailing whitespace in doc block
- Fix indentation in _is_valid_token (2-space -> 4-space)
- Fix inline comment spacing (PEP8: two spaces before #)
- Fix missing blank line before force_header_creds()
- Add changelog/66814.added.md
- Add tests/pytests/unit/netapi/cherrypy/test_header_creds.py with 12 unit tests
  covering force_header_creds(), salt_auth_tool(), and Events._is_valid_token()
